dbms/src/Storages/Transaction/KVStore.cpp

```diff
@@ -338,6 +338,15 @@ bool KVStore::tryFlushRegionData(UInt64 region_id, bool try_until_succeed, TMTCo
 {
     auto region_task_lock = region_manager.genRegionTaskLock(region_id);
     const RegionPtr curr_region_ptr = getRegion(region_id);
+    if (curr_region_ptr == nullptr)
+    {
+        /// If we can't find region here, we return true so proxy can trigger a CompactLog.
+        /// The triggered CompactLog will be handled by `handleUselessAdminRaftCmd`,
+        /// and result in a `EngineStoreApplyRes::NotFound`.
+        /// Proxy will print `region not found in engine-store, maybe have exec `RemoveNode` first`.
+        LOG_FMT_WARNING(log, "region {} [index: {}, term {}], not exist when flushing, maybe have exec `RemoveNode` first", region_id, index, term);
+        return true;
```

Comment:
If we return `false` here, can proxy skip the CompactLog and continue the `RemoveRegion`?

Reply:
In my understanding, the region is removed now in TiFlash which is exactly the state if the replay finishes.
However, I think we can both return false/true here:
- If we return false, the compact log is skipped, we alternatively forward a empty write to TiFlash which can be filtered by `NotFound` in TiFlash https://github.com/pingcap/tiflash/blob/2b8d3566531d12f4f33bdda4dbe9310f3202d478/dbms/src/Storages/Transaction/KVStore.cpp#L284
- If we return true, the compact log will be performed, but will also be filtered by `NotFound` in TiFlash https://github.com/pingcap/tiflash/blob/2b8d3566531d12f4f33bdda4dbe9310f3202d478/dbms/src/Storages/Transaction/KVStore.cpp#L404 **This is what we do before my refactoring**

Reply:
> the region is removed now in TiFlash which is exactly the state if the replay finishes

Here you assume that we get a nullptr because TiFlash removed Region X before that is done in proxy. So continue the compact log is OK.

Could there be a case that Region X is created in proxy and proxy calls tryFlushRegionData before the X is created in TiFlash?

Reply:
If such thing happens, previous and current code path will actually result in no-op on TiFlash' side.
However, I think such thing is not going to happen, and is safe even it do happens:
1. Regions are created in proxy by `replicate peer` and `BatchSplit`
2. For `replicate peer`, we must handle  a snapshot which contains region meta and data before we can handle raft logs. After the snapshot is handled, TiFlash certainly have corresponding region.
3. For `BatchSplit`, things are kind of complicated here. I think new regions should be created on TiFlash's side after we have executed BatchSplit as what is done in `handle_batch_split`. Otherwise it could be a bug.

However, the modification is when we want to flush a region before exec CompactLog, and there is no region on TiFlash, should we continue exec CompactLog. Actually the CompactLog can be considered as a dummy operation, no matter we return true or false here. If we don't have a region to flush/CompactLog, then it doesn't matter whether we have actually tried to do so.

Reply:
> If such thing happens, previous and current code path will actually result in no-op on TiFlash' side.
I agree that CompactLog is no-op on TiFlash side. My concern is whether that will cause the proxy to enter the wrong state.

It is OK if such thing is not going to happen